### ISSUE TITLE
Proposed improved API?

### ISSUE THREAD

**Issue by tzanio** [CEED/libCEED#8]
Below is a mockup of some proposed "improvements" in the current API. (This is all up for debate of course.)

Some of the rationale for these changes:
- The `Create` + `Set` + `Complete` workflow is self-documenting and flexible if we decide to introduce additional options in the future (e.g. a rectangular operator).
- `CeedBasis` separates dof and qpt descriptions, allows Lagrange elements e.g. in Chebyshev Lobatto points
- The `G` and `B` operators are not independent, so bundling them in a space makes sense
- `CeedOperator` is an abstraction for just operator action (as opposed to action and assembly). The `D` matrix (`qdata`) is stored internally in it.

@jedbrown, what do you think? Are you open to any of these changes?

```c
{
   // FE Basis = Operator B, B^t for a set of Eval modes
   CeedBasis basis;
   CeedBasisCreate(ceed, &basis);
   CeedBasisSetElement(basis, CEED_HEX); // implies 3D
   // Nodal scalar Lagrange basis, H1 or L2
   CeedBasisSetType(basis, CEED_LAGRANGE_BASIS, 3, CEED_GAUSS_LOBATTO);
   CeedBasisSetQuadrature(basis, 7, CEED_GAUSS); // order 7 = 4 points in 1D
   ierr = CeedBasisComplete(basis); // constructs B1d, etc., calls backend

   // Element restriction = Operator G, G^t
   CeedRestriction restriction;
   CeedRestrictionCreate(ceed, &restriction);
   CeedRestrictionSetElemSize(restriction, 64);
   CeedRestrictionSetNumElems(restriction, 8);
   CeedRestrictionSetNumDofs(restriction, 343);
   CeedInt *element_indices;
   // Set element_indices from mesh topology
   CeedRestrictionSetIndexType(restriction, element_indices,
                               CEED_MEM_HOST, CEED_USE_POINTER); // calls backend
   ierr = CeedRestrictionComplete(restriction); // error handling

   // Isoparametric case, otherwise define mesh-specific restriction & basis
   CeedSpace mesh_space;
   CeedSpaceCreate(ceed, &mesh_space);
   CeedSpaceSetRestriction(mesh_space, &restriction);
   CeedSpaceSetBasis(mesh_space, &basis);
   CeedSpaceSetNumComponents(mesh_space, 3);
   ierr = CeedSpaceComplete(mesh_space); // can call backend
   CeedVector mesh_nodes;
   // Set mesh nodes...

   // FE Space = Operators G, B, ...
   CeedSpace space;
   CeedSpaceCreate(ceed, &space);
   CeedSpaceSetRestriction(space, &restriction);
   CeedSpaceSetBasis(space, &basis);
   ierr = CeedSpaceComplete(space); // can call backend

   // FE Operator = Combines the mesh, the space, and kernels for assembly and
   // action of Operator D
   CeedOperator poisson;
   CeedOperatorCreate(ceed, &poisson);
   CeedOperatorSetMesh(poisson, mesh_space, mesh_nodes);
   CeedOperatorSetSpace(poisson, space);

   // Default kernels (set of Q-functions provided by libCEED)
   CeedOperatorSetKernels(poisson, "poisson");

   // Alternative: Custom kernels based on Q-functions
   CeedQFunction qf_poisson3d, qf_buildcoeffs;
   CeedQFunctionCreateInterior(ceed, 8, 1, 10*sizeof(CeedScalar), CEED_EVAL_GRAD,
                               CEED_EVAL_GRAD, f_poisson3d, "ex1.c:f_poisson3d",
                               &qf_poisson3d);
   CeedQFunctionCreateInterior(ceed, 1, 3, 10*sizeof(CeedScalar),
                               CEED_EVAL_INTERP | CEED_EVAL_GRAD, CEED_EVAL_NONE,
                               f_buildcoeffs,
                               "ex1.c:f_buildcoeffs", &qf_buildcoeffs);
   CeedOperatorSetAssemblyKernel(poisson, qf_my_poisson3d_assemble);
   CeedOperatorSetActionKernel(poisson, qf_my_poisson3d_apply);

   ierr = CeedOperatorComplete(poisson); // can call backend

   ierr = CeedOperatorSetup(poisson, CEED_REQUEST_IMMEDIATE); // calls backend

   ierr = CeedOperatorApply(poisson, u, r, CEED_REQUEST_IMMEDIATE); // calls backend
}
```

**Comment by jeremylt** [comment]
Closed by PR #41 (I'm pretty sure)

**Comment by v-dobrev** [comment]
@jeremylt this is an old discussion with suggestions for API changes that #41 does not address.

**Comment by jedbrown** [comment]
I think too much of this discussion is obsolete given API changes that have occurred since. Can you create a new issue for any outstanding concerns that you would like to address?

**Comment by v-dobrev** [comment]
Most of the example above is still valid. The only thing that #41 changes is how the "assemble" and "apply" kernels (actually called "operators" in the master) are constructed - if constructed explicitly. The above example proposes (among other things) a way to access operators already defined by libCEED, e.g.
```c
  CeedOperatorSetKernels(poisson, "poisson");
```
This type of functionality is something we still need to add.

**Comment by jedbrown** [comment]
See #37 ?

**Comment by v-dobrev** [comment]
That is just one of the things proposed in the example above.

**Comment by jedbrown** [comment]
This thread has too many non-orthogonal suggestions and the discussion is too biased by the most disruptive of those, which was addressed in #41. Please make new issues at a more modest granularity, as has been done for #37. (You have been largely silent in this project for the past few months. We welcome your comments, but this thread is too broad in scope and too stale for productive conversation.)
